### PR TITLE
RPE trajectory

### LINKED ISSUE #10: RPE trajectory
Create trajectory for (Golden) RPE including Johannes' sunflower trajectory

Check list:
- [x]     Create branch
- [ ]     Implement code
- [ ]     Add comments
- [ ]     Implement tests
- [ ]     Squash commmits
- [ ]     Create PR
- [ ]     Assign reviewer(s) to PR

### PR DESCRIPTION
**Pull request by ckolbPTB**
closes #10

### CODE REVIEW

File: src/mrpro/data/_KTrajectory.py [hunk 1]
```diff
@@ -37,6 +37,225 @@ def calc_traj(self, header: KHeader) -> torch.Tensor:
         ...
 
 
+class KTrajectoryRpe(KTrajectory):
```
Comment (schuenke) [inline src/mrpro/data/_KTrajectory.py]: I would suggest to move KTrajectoryRpe and KTrajectorySunflowerGoldenRpe into one or even two separate file(s). We will have several other trajectories and implementing them all in a single file might become messy.

File: src/mrpro/data/_KTrajectory.py [hunk 2]
```diff
@@ -37,6 +37,225 @@ def calc_traj(self, header: KHeader) -> torch.Tensor:
...
+            MR raw data header (KHeader) containing required meta data
+        """
+        kang = kang.flatten()
+        _, indices = np.unique(kang, return_index=True)
+        ang_idx = np.argsort([kang[index] for index in sorted(indices)])
+        shift_idx = np.zeros(max(ang_idx) + 1)
+        shift_idx[ang_idx] = np.arange(len(ang_idx))
+
+        # Apply sunflower shift
+        GoldenCut = 0.5 * (np.sqrt(5) + 1)
```
Comment (schuenke) [inline src/mrpro/data/_KTrajectory.py]: _GoldenCut_ should be all lower case and renamed to _golden_cut_ etc.

File: src/mrpro/data/_KTrajectory.py [hunk 3]
```diff
@@ -37,6 +37,225 @@ def calc_traj(self, header: KHeader) -> torch.Tensor:
...
+    """
+
+    def __init__(
+        self,
+        angle: float,
+        shift_between_rpe_lines: torch.Tensor = torch.tensor([0, 2, 1, 3]),
+        rad_us_factor: float = 1.0,
+    ) -> None:
+        super().__init__()
+        self.angle: float = angle
```
Comment (schuenke) [inline src/mrpro/data/_KTrajectory.py]: does it make sense to limit the angle between 0 and 2pi using `self.angle: float = angle % (2 * np.pi)` ?

File: src/mrpro/data/_KTrajectory.py [hunk 4]
```diff
@@ -37,6 +37,225 @@ def calc_traj(self, header: KHeader) -> torch.Tensor:
...
+    https://doi.org/10.1118/1.4890095 (open access).
+
+    Parameters
+    ----------
+    angle
+        angle in rad between two radial phase encoding lines
+    shift_between_rpe_lines
+        shift between radial phase encoding lines along the radial direction.
+        See _apply_shifts_between_rpe_lines() for more details
+    rad_us_factor
```
Comment (schuenke) [inline src/mrpro/data/_KTrajectory.py]: the unsersampling factor (`rad_us_factor`) is currently not used. Is this on purpose? Is it still work in progress? If we don't use it yet, I would vote to remove it to prevent confusion.

File: src/mrpro/data/_KTrajectory.py [hunk 5]
```diff
@@ -37,6 +37,225 @@ def calc_traj(self, header: KHeader) -> torch.Tensor:
...
+
+        # Angles of phase encoding lines
+        kang = kheader.acq_info.idx.k2 * self.angle
+
+        # K-space locations along phase encoding lines
+        krad = (kheader.acq_info.idx.k1 - kheader.encoding_limits.k1.center).to(torch.float32)
+        krad = self._apply_shifts_between_rpe_lines(krad, (kheader.acq_info.idx.k2 - kheader.encoding_limits.k2.center))
+        krad *= 2 * torch.pi / kheader.encoding_limits.k1.max
+
+        return self._combine_to_3d_traj(krad, kang, k0)
```
Comment (schuenke) [inline src/mrpro/data/_KTrajectory.py]: Would it make sense to store the calculated trajectory in our KTrajectory objects as well? Or should it be returned only? I was thinking about simple KTrajectory.plot() methods and currently such visualisation and/or other further processing methods would require to rerun the `calc_traj()` method

File: tests/data/test_ktraj.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+def rpe_traj_shape(valid_rpe_kheader):
+    """Expected shape of trajectory based on KHeader."""
+    nk0 = valid_rpe_kheader.acq_info.number_of_samples[0, 0, 0]
+    nk1 = valid_rpe_kheader.acq_info.idx.k1.shape[2]
+    nk2 = valid_rpe_kheader.acq_info.idx.k1.shape[1]
+    return torch.Size([1, 3, nk2, nk1, nk0])
+
+
+def test_KTrajectoryRpe(valid_rpe_kheader):
```
Comment (schuenke) [inline tests/data/test_ktraj.py]: I still propose that we test all class methods and also as many initialization cases as possible. For KTrajectoryRpe I thus propose to introduce different tests that cover for example the shape, different `angle `between the spokes, different `shift_between_rpe_lines`, different `rad_us_factor` (if we keep it) etc...

This could look like this:

```python
def test_KTrajectoryRpe_shape(valid_rpe_kheader):
    """Calculate RPE trajectory."""
    ktrajectory = KTrajectoryRpe(angle=np.pi * 0.618034)
    ktraj = ktrajectory.calc_traj(valid_rpe_kheader)
    assert ktraj.shape == rpe_traj_shape(valid_rpe_kheader)


def test_KTrajectoryRpe_angle(valid_rpe_kheader):
    ktrajectory1 = KTrajectoryRpe(angle=np.pi * 0.618034)
    ktrajectory2 = KTrajectoryRpe(angle=np.pi * 0.618034 + 2 * np.pi)
    ktraj1 = ktrajectory1.calc_traj(valid_rpe_kheader)
    ktraj2 = ktrajectory2.calc_traj(valid_rpe_kheader)
    assert np.array_equal(ktraj1, ktraj2)

def test_KTrajectoryRpe_rad_us_factor(valid_rpe_kheader):
    ...

```

File: tests/data/test_ktraj.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    """KHeader with all necessary parameters for RPE trajectories."""
+    # K-space dimensions
+    nk0 = 200
+    nk1 = 20
+    nk2 = 10
+
+    # List of k1 and k2 indices in the shape (d4, k2, k1)
+    k1 = torch.linspace(0, nk1 - 1, nk1, dtype=torch.int32)
+    k2 = torch.linspace(0, nk2 - 1, nk2, dtype=torch.int32)
+    idx_k1, idx_k2 = torch.meshgrid(k1, k2)
```
Comment (schuenke) [inline tests/data/test_ktraj.py]: This should be `idx_k1, idx_k2 = torch.meshgrid(k1, k2, indexing='ij')` to avoid Warning:

**_UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at ..\aten\src\ATen\native\TensorShape.cpp:3484.)_**

**Review by schuenke — COMMENTED** [review]
Please see my comments...

File: src/mrpro/data/_KTrajectory.py (hunk 5 above)
Comment (fzimmermann89) [reply src/mrpro/data/_KTrajectory.py]: No, the trajectory is not meant as a data storing class, it is only used to calculate trajectories. All stored state shall be settings for the trajectory. Not a concrete Tensor of kspace points.
As this can be different for different aquisitions.

The trajectory is saved only in Kdata.traj

So if we want to have a lot function, it should take in a Kdata object.

File: src/mrpro/data/_KTrajectory.py (hunk 3 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KTrajectory.py]: I would not pose any restrictions on the angle because the trajectory calculation has to be the same as in the sequence. Although I cannot give a good reason why somebody would use _angle > 2 * pi_ it might be needed for something and then we should not change this here.

File: tests/data/test_ktraj.py (hunk 6 above)
Comment (ckolbPTB) [reply tests/data/test_ktraj.py]: I will try to add some more tests. 

fyi "test_KTrajectoryRpe_angle" fails because pytorch does not seem to be too accurate. For me this
`torch.cos(torch.tensor(np.pi * 0.618034)) - torch.cos(torch.tensor(np.pi * 0.618034 + 2 * np.pi))`
yields `tensor(-5.9605e-08)` which would be fine when comparing with `almost_equal` but for 100 spokes: 
`torch.cos(100*torch.tensor(np.pi * 0.618034)) - torch.cos(100*torch.tensor(np.pi * 0.618034 + 2 * np.pi))` 
yields `tensor(1.0848e-05)`

`np.cos(np.pi * 0.618034) - np.cos(np.pi * 0.618034 + 2 * np.pi)` yields
`-4.996003610813204e-16`

Comment (schuenke) [reply tests/data/test_ktraj.py]: When I adjust the init to `self.angle: float = angle % (2 * np.pi)` (see comment above), this test actually passes even using `array_equal()`

File: src/mrpro/data/_KTrajectory.py (hunk 4 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KTrajectory.py]: left over from an old version -> removed

File: tests/data/test_ktraj.py (hunk 6 above)
Comment (ckolbPTB) [reply tests/data/test_ktraj.py]: yes, but if somebody used `torch.tensor(np.pi * 0.618034 + 2 * np.pi)` to calculate the angle in pypulseq then the trajectory would be wrong in mrpro if we used `self.angle: float = angle % (2 * np.pi)`

File: tests/data/test_ktraj.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+def rpe_traj_shape(valid_rpe_kheader):
+    """Expected shape of trajectory based on KHeader."""
+    nk0 = valid_rpe_kheader.acq_info.number_of_samples[0, 0, 0]
+    nk1 = valid_rpe_kheader.acq_info.idx.k1.shape[2]
+    nk2 = valid_rpe_kheader.acq_info.idx.k1.shape[1]
+    return torch.Size([1, 3, nk2, nk1, nk0])
+
+
+def test_KTrajectoryRpe_uniform(valid_rpe_kheader):
```
Comment (schuenke) [inline tests/data/test_ktraj.py]: Is it worth testing the shape of both, a uniform and a golden angle distribution? There is no reason why the shapes should differ, is there?

I think it would make more sense to test for specific implementation details, for example assert that the k-space center is never shifted (e.g. `torch.all(ktraj[0, 0, :, :, center_idx] == 0)`

Comment (ckolbPTB) [reply tests/data/test_ktraj.py]: I modified this test to actually check ktraj pos for angle_gap and angle_gap//2 for matching RPE lines. (actually found a bug in the test code in the process  :) )

File: src/mrpro/data/_KTrajectoryRpe.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        angle in rad between two radial phase encoding lines
+    shift_between_rpe_lines
+        shift between radial phase encoding lines along the radial direction.
+        See _apply_shifts_between_rpe_lines() for more details
+    """
+
+    def __init__(
+        self,
+        angle: float,
+        shift_between_rpe_lines: torch.Tensor = torch.tensor([0, 2, 1, 3]),
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectoryRpe.py]: We could consider making this a tuple of ints?
This can only be int, only 1d, never differentiable, right?

File: src/mrpro/data/_KTrajectoryRpe.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+        Parameters
+        ----------
+        krad
+            k-space positions along each phase encoding line
+        kang_idx
+            indices of angles to be used for shift calculation
+        """
+        delta_shift = 1 / len(self.shift_between_rpe_lines)
+        for ind in range(len(self.shift_between_rpe_lines)):
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectoryRpe.py]: for ind, shift in enumerate(self.shift_between_rpe_lines)

File: src/mrpro/data/_KTrajectoryRpe.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+            curr_angle_idx = torch.nonzero(
+                torch.fmod(kang_idx, len(self.shift_between_rpe_lines)) == ind, as_tuple=True
+            )
+            curr_shift = self.shift_between_rpe_lines[ind] * delta_shift
+            curr_krad = krad[curr_angle_idx]
+
+            # Do not shift the k-space center
+            center_idx = curr_krad == 0
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectoryRpe.py]: ==0 or approximately equal 0?

Comment (fzimmermann89) [reply src/mrpro/data/_KTrajectoryRpe.py]: curr_krad +=  curr_shift*(curr_krad!=0) ?

File: tests/data/test_ktraj.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    """KHeader with all necessary parameters for RPE trajectories."""
+    # K-space dimensions
+    nk0 = 200
+    nk1 = 20
+    nk2 = 10
+
+    # List of k1 and k2 indices in the shape (d4, k2, k1)
+    k1 = torch.linspace(0, nk1 - 1, nk1, dtype=torch.int32)
+    k2 = torch.linspace(0, nk2 - 1, nk2, dtype=torch.int32)
+    idx_k2, idx_k1 = torch.meshgrid(k2, k1, indexing='ij')
```
Comment (fzimmermann89) [inline tests/data/test_ktraj.py]: consider `idx_k1, idx_k2 = torch.meshgrid(k1, k2, indexing='xy')`, as xy will become default in pytorch

File: src/mrpro/data/_KTrajectoryRpe.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+            k-space points along readoud
+
+        Returns
+        -------
+            3D k-space trajectory
+        """
+        k1 = torch.repeat_interleave((krad * torch.cos(kang))[..., None], k0.shape[0], dim=3)
+        k2 = torch.repeat_interleave((krad * torch.sin(kang))[..., None], k0.shape[0], dim=3)
+        k0 = k0.expand(k1.shape)
+        return torch.concatenate([k0[:, None, :, :, :], k1[:, None, :, :, :], k2[:, None, :, :, :]], dim=1)
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectoryRpe.py]: For comparison, an einops way would be:
```
k1=einops.repeat(krad * torch.cos(kang),"...->... k0",k0=k0.shape[-1])
k2=einops.repeat(krad * torch.sin(kang),"...->... k0",k0=k0.shape[-1])
k0 = k0.expand(k1.shape)
return einops.rearrange([k0,k1,k2],"dir other k2 k1 k0->other dir k2 k1 k0")
```
Which to me is easier to understand,..

File: src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        kang
+            angles of the radial phase encoding lines
+        kheader
+            MR raw data header (KHeader) containing required meta data
+        """
+        kang = kang.flatten()
+        _, indices = np.unique(kang, return_index=True)
+        ang_idx = np.argsort([kang[index] for index in sorted(indices)])
+        shift_idx = np.zeros(max(ang_idx) + 1)
+        shift_idx[ang_idx] = np.arange(len(ang_idx))
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py]: as np.unique (and torch.unique with default sorted=True) return sorted values,
`shift_idx = np.argsort(indices)` or am I missing some case? If I am wrong, please add a comment why it has to be this complicated...
Also, is there a reason to first allocate a numpy zero-array and then fill it?

File: src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+        # Apply sunflower shift
+        golden_cut = 0.5 * (np.sqrt(5) + 1)
+        for ind in range(int(len(shift_idx))):
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py]:
```  
      for ind, shift in enumerate(shift_idx):
            krad[kheader.acq_info.idx.k2 == ind] +=((shift * golden_cut) % 1) - 0.5
 ```    
        
does the same.
        
Please check if the braces for `*2*0.5` are correct.
          
  Or is there a numerical reason why it has to be done that way? If yes, consider adding a comment for future reference

File: src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+            MR raw data header (KHeader) containing required meta data
+        """
+        kang = kang.flatten()
+        _, indices = np.unique(kang, return_index=True)
+        ang_idx = np.argsort([kang[index] for index in sorted(indices)])
+        shift_idx = np.zeros(max(ang_idx) + 1)
+        shift_idx[ang_idx] = np.arange(len(ang_idx))
+
+        # Apply sunflower shift
+        golden_cut = 0.5 * (np.sqrt(5) + 1)
```
Comment (fzimmermann89) [inline src/mrpro/data/_KTrajectorySunflowerGoldenRpe.py]: This is the constant phi, aka golden_ratio , correct?
Maybe consider naming it golden_ratio or phi?
(Why is it named "cut" )

**Review by fzimmermann89 — COMMENTED** [review]
some small comments and suggestions. Not sure I completely understood all of the logic. If I misunderstood it, you might consider adding some comments.

File: src/mrpro/data/_KTrajectoryRpe.py (hunk 9 above)
Comment (ckolbPTB) [reply src/mrpro/data/_KTrajectoryRpe.py]: tuple[float,...] because it does not have to be integer

Comment (fzimmermann89) [reply src/mrpro/data/_KTrajectoryRpe.py]: If it is float, I'd expect to specify normalized values (0, 0.5, 0.25, 0.75).
The normalizatoin by the length confused me and made me think it has to be integer.
If it now can be float, we might consider ignoring my comment and leaving it as a tensor to be differentiable in the offset?

Comment (ckolbPTB) [reply src/mrpro/data/_KTrajectoryRpe.py]: I changed it to normalized values as input rather than normalising with 1/length later

**Comment by schuenke** [comment]
When we planned the package structure, the idea was that the "data" subfolder should contain all data classes (e.g., KData, IData, QData, CsmData, DcfData...). Should we move everything related to trajectories into a mrpro.traj subfolder or would you like it to stay in the data folder?

**Review by fzimmermann89 — APPROVED** [review]
